## 1.3.1 Procedures as Arguments

### Example 1: summation

Consider the following three procedures.

The first computes the sum of the integers from a through b:

The second computes the sum of the cubes of the integers in the given range:

The third computes the sum of a sequence of terms in the series  

$$\frac{1}{1 \centerdot 3} + \frac{1}{5 \centerdot 7} + \frac{1}{9 \centerdot 11} + \ldots$$

which converges to $\pi$/8 (very slowly):

Indeed, mathematicians long ago identified the abstraction of summation of a series and invented "sigma notation", for example,to express this concept.

$$\sum_{n=a}^bf(n) = f(a) + \ldots+ f(b)$$

With the aid of the procedures sum as follows,we can express the mathematical concepts of summation:

So,we can redefine the process sum-integers to compute the sum of the integers from a through b:

In [1]:
cat 1.3/sum_integers_for_example.scm

(load "1.3/sum_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


 Redefine the process sum-cubes to compute the sum of the cubes of the integers from a through b:

In [2]:
cat 1.3/sum_cubes_for_example.scm

(load "1.3/sum_by_recursion.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


We can also redefine pi-sum in the same way:

In [5]:
cat 1.3/sum_pi_by_recursion.scm

(load "1.3/sum_by_recursion.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Example 2: numerical integration

Once we have sum, we can use it as a building block in formulating further concepts. For instance, the definite integral of a function f between the limits a and b can be approximated numerically using the formula

$$\int_a^bf=\left[f\left(a + \frac{dx}{2}\right) + f\left(a + dx + \frac{dx}{2}\right) + f\left(a + 2dx + \frac{dx}{2}\right) + \centerdot\centerdot\centerdot\right]dx$$

for small values of dx. We can express this directly as a procedure:

In [5]:
cat 1.3/compute_integral_for_function.scm

(load "1.3/sum_by_recursion.scm")

(define (integral f a b dx)
  (define (add-dx x) (+ x dx))
  (* (sum f (+ a (/ dx 2.0)) add-dx b) dx))


For example,we can let f is cube function:

### Exercise 1.29:
Simpson’s Rule is a more accurate method of numerical integration than the method illustrated above. Using Simpson’s Rule, the integral of a function f between a and b is approximated as

$$\frac{h}{3}(y_0 + 4y_1 + 2y_2 + 4y_3 + 2y_4 + \cdots + 2y_{n-2} + 4y_{n-1} + y_n)$$

where $h = (b -a)/n$,for some even integer n, and $y_k = f(a + kh)$.(Increasing n increases the accuracy of the approximation.)Define a procedure that takes as arguments $f$ , $a$, $b$, and $n$ and returns the value of the integral, computed using Simpson’s Rule.Use your procedure to integrate cube between 0 and 1 (with $n = 100$ and $n = 1000$), and compare the results to those of the integral procedure shown above.

### Exercise 1.30: 
The sum procedure above generates a linear recursion. The procedure can be rewrien so that the sum is performed iteratively. Show how to do this by filling in the missing expressions in the following definition:

### Answer:

In [1]:
cat 1.3/sum_by_iteration.scm

(define (sum term a next b)
  (define (iter a result)
    (if (> a b)
        result
        (iter (next a)
              (+ (term a) result))))
  (iter a 0))
		


1.Compute the sum of the integers from a to b:

In [2]:
cat 1.3/sum_integers_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.Compute the sum of the cubes of the integers from a to b:

In [3]:
cat 1.3/sum_cubes_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


3.Compute the sum-pi procedure by way of iterative process: 

In [4]:
cat 1.3/sum_pi_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Exercise 1.31:
a. The $sum$ procedure is only the simplest of a vast number of similar abstractions that can be captured as higher-order procedures.Write an analogous procedure called $product$ that returns the product of the values of a
function at points over a given range. Show how to define $factorial$ in terms of $product$. Also use product to compute approximations to $\pi$ using the formula

$$\frac{\pi}{4} = \frac{2\centerdot4\centerdot4\centerdot6\centerdot6\centerdot8\cdots}{3\centerdot3\centerdot5\centerdot5\centerdot7\centerdot7\cdots}$$

### Answer:

1.The product procedure as follows.

In [1]:
cat 1.3/product_by_recursion.scm

(define (product term a next b)
  (if (> a b) 1
      (* (term a)
         (product term (next a) next b))))


2.Redefine factorial procedure in terms of product procedure.

In [2]:
cat 1.3/factorial_by_product_by_recursion.scm

(load "1.3/product_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (factorial b)
  (product identity 1 inc b))


3.Use product to compute approximations to $\pi$.

In [3]:
cat 1.3/product_pi_by_recursion.scm

(load "1.3/product_by_recursion.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
  ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
  (product pi-term a pi-next b))


b. If your product procedure generates a recursive process,write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### Answer:

Since the product procedure above generates a recursive process;we should write one that generates an iterative process

1.The product procedure as follows.

In [5]:
cat 1.3/product_by_iteration.scm

(define (product term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a)
              (* (term a) result))))
  (iter a 1))


Test:

2.Redefine factorial procedure

In [6]:
cat 1.3/factorial_by_product_by_iteration.scm

(load "1.3/product_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (factorial b)
  (product identity 1 inc b))


3.Use product to compute approximations to $\pi$

In [7]:
cat 1.3/product_pi_by_iteration.scm

(load "1.3/product_by_iteration.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
  ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
  (product pi-term a pi-next b))


### Exercise 1.32:
#### a. 
Show that $sum$ and $product$ (Exercise 1.31) are both special cases of a still more general notion called $accumulate$ that combines a collection of terms, using some general accumulation function:

$accumulate$ takes as arguments the same term and range specifications as $sum$ and $product$, together with a $combiner$ procedure (of two arguments) that specifies how the current term is to be combined with the accumulation of the preceding terms and a null-value that specifies what base value to use when the terms run out. Write $accumulate$ and show how $sum$ and $product$ can both be defined as simple calls to $accumulate$.

1.The $accumulate$ recursive procedure:

In [8]:
cat 1.3/accumulate_by_recursion.scm

(define (accumulate combiner null-value term a next b)
  (if (> a b) null-value
      (combiner (term a)
                (accumulate combiner
                            null-value
                            term
                            (next a)
                            next
                            b))))


2.Redefine the $sum$ procedure by way of the $accumulate$ recursive procedure:

In [11]:
cat 1.3/sum_by_accumulate_by_recursion.scm

(load "1.3/accumulate_by_recursion.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


3.Redefine the $product$ procedure by way of the $accumulate$ recursive procedure:

In [14]:
cat 1.3/product_by_accumulate_by_recursion.scm

(load "1.3/accumulate_by_recursion.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))


#### Test:

1.summation

In [20]:
cat 1.3/sum_integers_by_accumulate_by_recursion.scm

(load "1.3/sum_by_accumulate_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.product

In [22]:
cat 1.3/product_pi_by_accumulate_by_recursion.scm

(load "1.3/product_by_accumulate_by_recursion.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
    (* 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
    (product pi-term a pi-next b))


#### b. 
If your $accumulate$ procedure generates a recursive process, write one that generates an iterative process.If it generates an iterative process, write one that generates a recursive process.

1.The accumulate iterative procedure:

In [10]:
cat 1.3/accumulate_by_iteration.scm

(define (accumulate combiner null-value term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a)
              (combiner (term a) result))))
  (iter a null-value))


2.Redefine the sum procedure by way of the accumulate iterative procedure:

In [13]:
cat 1.3/sum_by_accumulate_by_iteration.scm

(load "1.3/accumulate_by_iteration.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


3.Redefine the $product$ procedure by way of the $accumulate$ iterative procedure:

In [15]:
cat 1.3/product_by_accumulate_by_iteration.scm

(load "1.3/accumulate_by_iteration.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))


#### Test:

1.summation

In [21]:
cat 1.3/sum_integers_by_accumulate_by_iteration.scm

(load "1.3/sum_by_accumulate_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.product

In [23]:
cat 1.3/product_pi_by_accumulate_by_iteration.scm

(load "1.3/product_by_accumulate_by_iteration.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
    ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
    (product pi-term a pi-next b))


### Exercise 1.33:
You can obtain an even more general version of $accumulate$ (Exercise 1.32) by introducing the notion of a $filter$ on the terms to be combined. That is, combine only those terms derived from values in the range that satisfy a specified condition. The resulting $filtered$-$accumulate$ abstraction takes the same arguments as $accumulate$, together with an additional predicate of one argument that specifies the $filter$. Write filtered accumulate as a procedure.Show how to express the following using $filtered$-$accumulate$:

### Answer:

#### A.The filtered-accumulate recursive procedure:

In [24]:
cat 1.3/filtered_accumulate_by_recursion.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (if (> a b) null-value
      (if (valid a)
         (combine (term a)
                  (filtered-accumulate combine
                                       null-value
                                       term
                                       (next a)
                                       next
                                       b
                                       valid))
        (filtered-accumulate combine
       	                     null-value
       	                     term
       	                     (next a)
       	                     next
       	                     b
       	                     valid))))


#### a. 
the $sum$ of the squares of the prime numbers in the interval a to b (assuming that you have a $prime$ predicate already written)

In [25]:
cat 1.3/sum_square_of_prime_by_filtered_accumulate_by_recursion.scm

(load "1.3/filtered_accumulate_by_recursion.scm")
(load "1.2/prime_by_search_divisor1.scm")

(define (inc n) (+ n 1))

(define (sum-prime a b)
  (filtered-accumulate + 0 square a inc b prime))


#### b. 
the $product$ of all the positive integers less than n that are relatively prime to n  
(i.e., all positive integers $i < n$ such that $GCD(i,n) = 1$).

In [26]:
cat 1.3/product_of_coprimes.scm

(load "1.3/filtered_accumulate_by_recursion.scm")
(load "1.2/gcd_by_Euclid_Algorithm.scm")

(define (identity x) x)
(define (coprime a b) (and (< a b) (= 1 (gcd a b))))
(define (inc n) (+ n 1))

(define (product-of-coprimes n)
  (filtered-accumulate * 1 identity 1 inc n (lambda (x) (coprime x n))))


#### B.The filtered-accumulate iterative procedure:

In [1]:
cat 1.3/filtered_accumulate_by_iteration.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (define (iter i result)
    (cond ((> i b) result)
          ((valid i)
           (iter (next i) (combine (term i) result)))
          (else (iter (next i) result))))
  (iter a null-value))


#### a.
the sum of the cubes of the prime numbers in the interval a to b:

In [2]:
cat 1.3/sum_cube_of_prime_by_filtered_accumulate_by_iteration.scm

(load "1.3/filtered_accumulate_by_iteration.scm")
(load "1.2/prime_by_search_divisor1.scm")

(define (inc n) (+ n 1))
(define (cube x) (* x x x))

(define (sum-prime a b)
  (filtered-accumulate + 0 cube a inc b prime))


#### b.
the product of all the positive even integers less than n to n:

In [3]:
cat 1.3/sum_even_integers_by_filtered_accumulate_by_iteration.scm

(load "1.3/filtered_accumulate_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-even-integers n)
  (filtered-accumulate + 0 identity 1 inc n even?))


## 1.3.2 Constructing Procedures Using $lambda$

In using $sum$ as in Section 1.3.1, it seems terribly awkward to have to define trivial procedures such as $pi$-$term$ and $pi$-$next$ just so we can use them as arguments to our higher-order procedure. Rather than define $pi$-$next$ and $pi$-$term$, it would be more convenient to have a way to directly specify “the procedure that returns its input incremented by 4” and “the procedure that returns the reciprocal of its input times its input plus 2.” We can do this by introducing the special form $lambda$, which creates procedures. Using $lambda$ we can describe what we want as

and

Then our $pi$-$sum$ procedure can be expressed without defining any auxiliary procedures as

In [4]:
cat 1.3/sum_pi_by_lambda_by_recursion.scm

(load "1.3/sum_by_recursion.scm")

(define (pi-sum a b)
  (sum (lambda (x) (/ 1.0 (* x (+ x 2))))
       a
       (lambda (x) (+ x 4))
       b))


Again using $lambda$, we can write the $integral$ procedure without having to define the auxiliary procedure $add$-$dx$:

In [6]:
cat 1.3/compute_integral_for_function_by_lambda.scm

(load "1.3/sum_by_recursion.scm")

(define (integral f a b dx)
  (* (sum f
          (+ a (/ dx 2.0))
          (lambda (x) (+ x dx))
          b)
     dx))


In general, $lambda$ is used to create procedures in the same way as $define$, except that no name is specified for the procedure:

The resulting procedure is just as much a procedure as one that is created using $define$.   
The only difference is that it has not been associated with any name in the environment. In fact,

is equivalent to

Like any expression that has a procedure as its value, a $lambda$ expression can be used as the operator in a combination such as

### Using $let$ to create $local$ $variables$

Another use of $lambda$ is in creating $local$ $variables$.   
We often need $local$ $variables$ in our procedures other than those that have been bound as $formal$ $parameters$.   
For example, suppose we wish to compute the function

$$f(x,y) = x(1 + xy)^2 + y(1 - y) + (1 + xy)(1 -y),$$

which we could also express as

$$a = 1 + xy,$$
$$b = 1 - y,$$
$$f(x,y) = xa^2 + yb + ab.$$

In writing a procedure to compute $f$ , we would like to include as $local$ $variables$ not only $x$ and $y$ but also the names of intermediate quantities like $a$ and $b$. One way to accomplish this is to use an auxiliary procedure to bind the $local$ $variables$:

Of course, we could use a $lambda$ expression to specify an anonymous procedure for binding our $local$ $variables$. The body of $f$ then becomes a single call to that procedure:

This construct is so useful that there is a special form called $let$ to make its use more convenient.  
Using $let$, the $f$ procedure could be written as

The general form of a $let$ expression is

The first part of the $let$ expression is a list of $name$-$expression$ pairs.  
When the $let$ is evaluated, each $name$ is associated with the value of the corresponding $expression$.   
The body of the $let$ is evaluated with these names bound as $local$ $variables$.   
The way this happens is that the $let$ expression is interpreted as an alternate syntax for

No new mechanism is required in the interpreter in order to provide $local$ $variables$.   
A $let$ expression is simply $syntactic$ $sugar$ for the underlying $lambda$ application

We can see from this equivalence that the scope of a variable specified by a $let$ expression is the body of the $let$.   
This implies that:

$\bullet$ $let$ allows one to bind variables as locally as possible to where they are to be used. For example, if the value of $x$ is 5, the value of the expression

is 38. Here, the $x$ in the body of the $let$ is 3, so the value of the let expression is 33.   
On the other hand, the $x$ that is the second argument to the outermost $+$ is still 5.

$\bullet$ The variables’values are computed outside the $let$. This matters when the expressions that provide the values for the $local$ $variables$ depend upon variables having the same names as the local variables themselves. For example, if the value of $x$ is 2, the expression

will have the value 12 because, inside the body of the $let$, $x$ will be 3 and $y$ will be 4 (which is the outer $x$ plus 2).

Sometimes we can use internal definitions to get the same effect as with $let$. For example, we could have defined the procedure $f$ above as

We prefer,however,to use $let$ in situations like this and to use $internal$ $define$ only for $internal$ $procedures$.

### Exercise 1.34: 
Suppose we define the procedure

Then we have

What happens if we (perversely) ask the interpreter to evaluate the combination $(f$ $f)$? Explain.

## 1.3.3 Procedures as General Methods

### Example 3: Finding roots of equations by the half-interval method

The $half$-$interval$ $method$ is a simple but powerful technique for finding roots of an equation $f(x) = 0$, where $f$ is a continuous function. The idea is that, if we are given points $a$ and $b$ such that $f(a) < 0 < f(b)$,then $f$ must have at least one zero between $a$ and $b$. To locate a zero,let $x$ be the average of $a$ and $b$, and compute $f(x)$. If $f(x) > 0$, then $f$ must have a zero between $a$ and $x$. If $f(x) < 0$, then $f$ must have a zero between $x$ and $b$. Continuing in this way, we can identify smaller and smaller intervals on which $f$ must have a zero. When we reach a point where the interval is small enough, the process stops. Since the interval of uncertainty is reduced by half at each step of the process, the number of steps required grows as $\Theta(log(L/T))$, where $L$ is the length of the original interval and $T$ is the error tolerance (that is, the size of the interval we will consider “small enough”). Here is a procedure that implements this strategy:

$search$ is awkward to use directly, because we can accidentally give it points at which $f$’s values do not have the required sign, in which case we get a wrong answer. Instead we will use search via the following procedure, which checks to see which of the endpoints has a negative function value and which has a positive value, and calls the search procedure accordingly. If the function has the same sign on the two given points, the half-interval method cannot be used, in which case the procedure signals an error.

We can group the processes above:

In [1]:
cat 1.3/half_interval_method.scm

(define (average a b) (/ (+ a b) 2.0))
(define (close-enough x y) (< (abs (- x y)) 0.001))

(define (search f neg-point pos-point)
  (let ((midpoint (average neg-point pos-point)))
    (if (close-enough neg-point pos-point) midpoint
      (let ((test-value (f midpoint)))
        (cond ((positive? test-value) (search f neg-point midpoint))
              ((negative? test-value) (search f midpoint pos-point))
              (else midpoint))))))

(define (half-interval-method f a b)
  (let ((a-value (f a)) (b-value (f b)))
    (cond ((and (negative? a-value) (positive? b-value)) (search f a b))
          ((and (negative? b-value) (positive? a-value)) (search f b a))
          (else (error "Values are not of opposite sign" a b)))))


The following example uses the $half$-$interval$ $method$ to approximate $\pi$ as the root between 2 and 4 of $sin x = 0$:

Here is another example, using the $half$-$interval$ $method$ to search for a root of the equation $x^3 - 2x - 3 = 0$ between 1 and 2:

### Example 4: Finding fixed points of functions

A number $x$ is called a $fixed$ $point$ of a function $f$ if $x$ satisfies the equation $f(x) = x$. For some functions $f$ we can locate a $fixed$ $point$ by beginning with an initial $guess$ and applying $f$ repeatedly,

$$f(x),\;\;\;f(f(x)),\;\;\;f(f(f(x))),\;\;\;\ldots,$$

until the value does not change very much. Using this idea, we can devise a procedure $fixed$-$point$ that takes as inputs a function and an initial guess and produces an approximation to a $fixed$ $point$ of the function. We apply the function repeatedly until we find two successive values whose difference is less than some prescribed tolerance:

In [1]:
cat 1.3/fixed_point_of_function.scm

(define tolerance 0.00001)
(define (fixed-point f first-guess)
  (define (close-enough v1 v2)
    (< (abs (- v1 v2)) tolerance))
  (define (try guess)
    (let ((next (f guess)))
      (if (close-enough guess next)
          next
          (try next))))
  (try first-guess))


For example, we can use this method to approximate the $fixed$ $point$ of the $cosine$ function, starting with 1 as an initial approximation:

Similarly, we can find a solution to the equation $y = siny + cosy$:

The $fixed$-$point$ process is reminiscent of the process we used for finding $square$ $roots$ in Section 1.1.7. Both are based on the idea of repeatedly improving a guess until the result satisfies some criterion. In fact, we can readily formulate the $square$-$root$ computation as a $fixed$-$point$ search. Computing the $square$ $root$ of some number $x$ requires finding a $y$ such that $y^2 = x$. Putting this equation into the equivalent form $y = x/y$,we recognize that we are looking for a fixed point of the function $y\mapsto x=y$, and we can therefore try to compute $square$ $roots$ as

Unfortunately, this $fixed$-$point$ $search$ does not converge. Consider an initial guess $y_1$. The next guess is $y_2 = x/y_1$ and the next guess is $y_3 = x/y_2 = x/(x/y_1) = y_1$. This results in an infinite loop in which the two guesses $y_1$ and $y_2$ repeat over and over, oscillating about the answer.

One way to control such oscillations is to prevent the guesses from changing so much. Since the answer is always between our guess $y$ and $x/y$, we can make a new guess that is not as far from $y$ as $x/y$ by averaging $y$ with $x/y$, so that the next guess after $y$ is $\frac{1}{2}(y + x/y)$ instead of $x/y$. The process of making such a sequence of guesses is simply the process of looking for a fixed point of $y\mapsto\frac{1}{2}(y + x/y)$:

(Note that $y = \frac{1}{2}(y + x/y)$ is a simple transformation of the equation $y = x/y$; to derive it, add y to both sides of the equation and divide by 2.)

With this modification, the $square$-$root$ procedure works. In fact, if we unravel the definitions, we can see that the sequence of approximations to the square root generated here is precisely the same as the one generated by our original $square$-$root$ procedure of Section 1.1.7. This approach of averaging successive approximations to a solution, a technique that we call $average$ $damping$, often aids the convergence of $fixed$-$point$ searches.

### Exercise 1.35: 
Show that the $golden$ $ratio$ $\varphi$ (Section 1.2.2) is a $fixed$ $point$ of the transformation $x \mapsto 1 + 1/x$, and use this fact to compute $\varphi$ by means of the $fixed$-$point$ procedure.

### Exercise 1.36: 
Modify $fixed$-$point$ so that it prints the sequence of approximations it generates, using the newline and display primitives shown in Exercise 1.22. Then find a solution to $x^x = 1000$ by finding a fixed point of $x \mapsto log(1000)/log(x)$. (Use Scheme’s primitive $log$ procedure, which computes natural logarithms.) Compare the number of steps this takes with and without $average$ $damping$. (Note that you cannot start $fixed$-$point$ with a guess of 1, as this would cause division by $log(1) = 0$.)

### Exercise 1.37:

a. An infinite $continued$ $fraction$ is an expression of the form

$$f = \cfrac{N_1}{D_1 + \cfrac{N_2}{D_2 + \cfrac{N_3}{D_3 + \cdots}}}$$

As an example, one can show that the infinite $continued$ $fraction$ expansion with the $N_i$ and the $D_i$ all equal to 1 produces $1/\varphi$, where $\varphi$ is the $golden$ $ratio$ (described in Section 1.2.2). One way to approximate an infinite continued fraction is to truncate the expansion after a given number of terms. Such a truncation—a so-called $k$-$term$ $finite$ $continued$ $fraction$—has the form

$$\cfrac{N_1}{D_1 + \cfrac{N_2}{\ddots + \cfrac{N_k}{D_k}}}.$$

Suppose that $n$ and $d$ are procedures of one argument (the term index $i$) that return the $N_i$ and $D_i$ of the terms of the continued fraction. Define a procedure $cont$-$frac$ such that evaluating ($cont$-$frac$ $n$ $d$ $k$) computes the value of the $k$-$term$ finite continued fraction.Check your procedure by approximating $1/\varphi$ using

for successive values of $k$. How large must you make $k$ in order to get an approximation that is accurate to 4 decimal places?

b. If your $cont$-$frac$ procedure generates a recursive process, write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### Exercise 1.38: 
In 1737, the Swiss mathematician Leonhard Euler published a memoir $De$ $Fractionibus$ $Continuis$, which included a continued fraction expansion for $e$ $-$ $2$, where $e$ is the base of the natural logarithms. In this fraction, the $N_i$ are all $1$, and the $D_i$ are successively $1, 2, 1, 1, 4, 1, 1, 6, 1, 1, 8, \ldots$. Write a program that uses your $cont$-$frac$ procedure from Exercise 1.37 to approximate $e$, based on Euler’s expansion.

### Exercise 1.39: 
A continued fraction representation of the $tangent$ function was published in 1770 by the German mathematician J.H. Lambert:

$$\tan x = \cfrac{x}{1 - \cfrac{x^2}{3 - \cfrac{x^2}{5 - \ldots}}}$$

where $x$ is in radians. Define a procedure ($tan$-$cf$ $x$ $k$) that computes an approximation to the $tangent$ function based on Lambert’s formula. $k$ specifies the number of terms to compute, as in Exercise 1.37.

## 1.3.4 Procedures as Returned Values

The above examples demonstrate how the ability to pass procedures as arguments significantly enhances the expressive power of our programming language. We can achieve even more expressive power by creating procedures whose returned values are themselves procedures.

We can illustrate this idea by looking again at the fixed-point example described at the end of Section 1.3.3. We formulated a new version of the square-root procedure as a fixed-point search, starting with the observation that $\sqrt{x}$ is a fixed-point of the function $y \mapsto x/y$.Then we used average damping to make the approximations converge. Average damping is a useful general technique in itself. Namely, given a function $f$, we consider the function whose value at $x$ is equal to the average of $x$ and $f(x)$.

We can express the idea of average damping by means of the following procedure:

$average$-$damp$ is a procedure that takes as its argument a procedure $f$ and returns as its value a procedure (produced by the lambda) that,when applied to a number x, produces the average of $x$ and $(f\;x)$. For example, applying $average$-$damp$ to the square procedure produces a procedure whose value at some number $x$ is the average of 
$x$ and $x^2$.Applying this resulting procedure to 10 returns the average of 10 and 100, or 55

Using $average$-$damp$, we can reformulate the square-root procedure as follows:

In [1]:
cat 1.3/square_root.scm

(load "1.3/fixed_point_of_function.scm")
(load "1.3/average_damp.scm")

(define (sqrt x)
  (fixed-point (average-damp (lambda (y) (/ x y))) 1.0))


Notice how this formulation makes explicit the three ideas in the method:fixed-point search, average damping, and the function $y \mapsto x/y$. It is instructive to compare this formulation of the square-root method with the original version given in Section 1.1.7. Bear in mind that these procedures express the same process, and notice how much clearer the idea becomes when we express the process in terms of these abstractions. In
general, there are many ways to formulate a process as a procedure. Experienced programmers know how to choose procedural formulations that are particularly perspicuous, and where useful elements of the process are exposed as separate entities that can be reused in other applications. As a simple example of $reuse$, notice that the cube root of $x$ is a fixed point of the function $y \mapsto x/y^2$, so we can immediately generalize our square-root procedure to one that extracts cube roots:

In [2]:
cat 1.3/cube_root.scm

(load "1.3/fixed_point_of_function.scm")
(load "1.3/average_damp.scm")

(define (cube-root x)
  (fixed-point (average-damp (lambda (y) (/ x (square y)))) 1.0))


### Example 5: Newton’s method

When we first introduced the square-root procedure, in Section 1.1.7, we mentioned that this was a special case of $Newton’s\; method$.If $x\mapsto g(x)$ is a differentiable function, then a solution of the equation $g(x)\,=\,0$ is a fixed point of the function $x \mapsto f(x)$, where

$$f(x) = x - \frac{g(x)}{Dg(x)}$$

and $Dg(x)$ is the derivative of $g$ evaluated at $x$.

Newton’s method is the use of the fixed-point method we saw above to approximate a solution of the equation by finding a fixed point of the function $f$.

For many functions $g$ and for sufficiently good initial guesses for $x$, Newton’s method converges very rapidly to a solution of $g(x)\, =\, 0$.

In order to implement Newton’s method as a procedure, we must first express the idea of derivative. Note that “derivative,” like average damping, is something that transforms a function into another function. For instance, the derivative of the function $x \mapsto x^3$ is the function $x \mapsto 3x^2$. In general, if $g$ is a function and dx is a small number, then the derivative $Dg$ of $g$ is the function whose value at any number $x$ is given (in the limit of small $dx$) by

$$Dg(x) = \frac{g(x + dx) - g(x)}{dx}.$$

Thus, we can express the idea of derivative (taking $dx$ to be, say, 0.00001) as the procedure

Like $average$-$damp$, $deriv$ is a procedure that takes a procedure as argument and returns a procedure as value. For example, to approximate the derivative of $x \mapsto x^3$ at 5 (whose exact value is 75) we can evaluate

With the aid of $deriv$, we can express Newton’s method as a fixed-point process:

The $newton$-$transform$ procedure expresses the formula at the beginning of this section, and newtons-method is readily defined in terms of this. It takes as arguments a procedure that computes the function for which we want to find a zero, together with an initial guess. For instance,to find the square root of x, we can use Newton’s method to find a zero of the function $y \mapsto y^2 -􀀀 x$ starting with an initial guess of 1.

This provides yet another form of the square-root procedure:

### Abstractions and first-class procedures

We’ve seen two ways to express the $square$-$root$ computation as an instance of a more general method, once as a $fixed$-$point\; search$ and once using $Newton’s\; method$. Since Newton’s method was itself expressed as a fixed-point process, we actually saw two ways to compute square roots as fixed points. Each method begins with a function and finds a fixed point of some transformation of the function. We can express this general idea itself as a procedure:

This very general procedure takes as its arguments a procedure $g$ that computes some function, a procedure that transforms $g$, and an initial guess. The returned result is a fixed point of the transformed function.

Using this abstraction, we can recast the first square-root computation from this section (where we look for a fixed point of the $average$-$damped$ version of $y \mapsto x/y$) as an instance of this general method:

Similarly, we can express the second square-root computation from this section (an instance of $Newton’s\; method$ that finds a fixed point of the $Newton\; transform$ of $y \mapsto y^2-x$) as